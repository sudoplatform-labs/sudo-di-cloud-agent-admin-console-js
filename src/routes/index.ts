export * from './route-definition';
export * from './utils';
