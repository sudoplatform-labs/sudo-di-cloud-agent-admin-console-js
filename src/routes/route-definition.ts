import { ImageName } from '../components/Image';
import { FeatureFlag } from '../feature-flags';

export interface RouteDefinition {
  /** Route path */
  path: string;

  /** Results in partial match if set to false (default = true) */
  exact?: boolean;

  /** Component to render for this route */
  component?: React.ComponentType;

  /** Child routes that are nested under this route */
  routes?: RouteDefinition[];

  /** Display name of route, used in Sidebar etc. */
  displayName?: string;

  /** Icon to be displayed in sidebar */
  iconName?: ImageName;

  /** Allow Icon classes to apply styling if necessary */
  iconClass?: string;

  /** Set to true if this route should appear in the sidebar */
  showInSidebar?: boolean;

  /** Restrict visibility based on a feature flag  */
  featureFlag?: FeatureFlag;
}

export type RouteDefinitionFactory = () => RouteDefinition;
