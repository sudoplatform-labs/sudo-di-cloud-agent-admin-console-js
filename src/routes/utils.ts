import { flatten, omit } from 'lodash';
import { matchPath } from 'react-router';
import { RouteDefinition } from './route-definition';

/** Flattens a heirarchy of routes into an array */
export function flattenRoutes(
  routes: RouteDefinition[],
  maxDepth?: number,
): RouteDefinition[] {
  const nextDepth = maxDepth && maxDepth - 1;

  return flatten(
    routes.map((route) => {
      const subRoutes =
        nextDepth === undefined || nextDepth > 1
          ? flattenRoutes(route.routes ?? [], nextDepth)
          : [];
      return [route, ...subRoutes];
    }),
  ).map((route) => omit(route, 'routes'));
}

/** Finds a route definition that matches a given path */
export function findRoute(
  urlPath: string,
  routes: RouteDefinition[],
): RouteDefinition | undefined {
  return routes.find(
    (route) =>
      route.exact !== false &&
      matchPath(urlPath, {
        path: route.path,
        exact: true,
      }),
  );
}

/** Returns all ancestor paths for `path`, including `path` itself */
export const getPathAncestry = (path?: string): string[] => {
  if (!path) return [];

  const pathSnippets = path.split('/').filter((snippet) => snippet);

  return pathSnippets.map((_, index) => {
    return `/${pathSnippets.slice(0, index + 1).join('/')}`;
  });
};
