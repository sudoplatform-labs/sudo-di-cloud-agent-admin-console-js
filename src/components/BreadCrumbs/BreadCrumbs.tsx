import { Breadcrumb } from 'antd';
import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { RouteDefinition, findRoute, getPathAncestry } from '../../routes';

export const StyledBreadCrumb = styled(Breadcrumb)`
  padding: 12px 24px;
`;

interface Props {
  urlPath: string;
  routes: RouteDefinition[];
}

interface ItemProps {
  path: string;
  linkText: string;
}

export const BreadCrumbs: React.FC<Props> = (props) => {
  const { urlPath } = props;
  const componentUrlPaths = getPathAncestry(urlPath);
  const items: ItemProps[] = componentUrlPaths
    .map((path) => {
      const route = findRoute(path, props.routes);
      return {
        path,
        linkText: route?.displayName ?? '',
      };
    })
    .filter((item) => !!item.linkText);

  return (
    <StyledBreadCrumb separator=">">
      {items.slice(0).map((itemProps) => (
        <Breadcrumb.Item key={itemProps.path}>
          <Link to={itemProps.path}>{itemProps.linkText}</Link>
        </Breadcrumb.Item>
      ))}
    </StyledBreadCrumb>
  );
};
