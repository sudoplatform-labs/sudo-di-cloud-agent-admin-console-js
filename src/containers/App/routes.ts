import { createRouteRedirect } from '../../components/RouteRenderer';
import { RouteDefinitionFactory } from '../../routes';
import { Console } from '../Console';

export const getRoot: RouteDefinitionFactory = () => ({
  path: '/',
  component: createRouteRedirect('/console'),
  routes: [
    {
      path: '/console',
      component: Console,
      exact: false,
    },
  ],
});
