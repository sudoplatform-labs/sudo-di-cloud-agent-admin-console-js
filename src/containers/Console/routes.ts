import { RouteDefinitionFactory } from '../../routes';
import * as Pages from '../../pages';
import { createRouteRedirect } from '../../components/RouteRenderer';

export const getConsoleRoutesRoot: RouteDefinitionFactory = () => ({
  path: '/console',
  component: createRouteRedirect('/console/dashboard'),
  routes: [
    {
      path: '/console/dashboard',
      component: Pages.ConsoleDashboard,
      displayName: 'Cloud Agent Dashboard',
      iconName: 'home',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/settings',
      component: Pages.Settings,
      displayName: 'Settings',
      iconName: 'settings',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/decentralized-identifiers',
      component: Pages.DecentralizedIdentifiersDashboard,
      displayName: 'Decentralized Identifiers',
      iconName: 'icon-did-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/connections',
      component: Pages.ConnectionsDashboard,
      displayName: 'Connections',
      iconName: 'icon-connection-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/holder-wallet',
      component: createRouteRedirect('/console/holder-wallet/credentials'),
      displayName: 'Holder Wallet',
      iconName: 'icon-wallet-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/holder-wallet/credentials',
          component: Pages.Credentials,
          displayName: 'My Credentials',
          showInSidebar: true,
        },
        {
          path: '/console/holder-wallet/proofs',
          component: Pages.PresentedProofs,
          displayName: 'My Proof Presentations',
          showInSidebar: true,
        },
      ],
    },
    {
      path: '/console/credential-issuer',
      component: createRouteRedirect('/console/credential-issuer/dashboard'),
      displayName: 'Credential Issuer',
      iconName: 'icon-issuer-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/credential-issuer/dashboard',
          component: Pages.CredentialIssuerDashboard,
          displayName: 'Dashboard',
          showInSidebar: true,
        },
        {
          path: '/console/credential-issuer/credential-catalogue',
          component: Pages.CredentialCatalogue,
          displayName: 'Catalogue',
          showInSidebar: true,
        },
        {
          path: '/console/credential-issuer/credential-requests',
          component: Pages.CredentialIssuance,
          displayName: 'Credential Issuance',
          showInSidebar: true,
        },
      ],
    },
    {
      path: '/console/credential-verifier',
      component: createRouteRedirect('/console/credential-verifier/dashboard'),
      displayName: 'Credential Verifier',
      iconName: 'icon-verifier-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/credential-verifier/dashboard',
          component: Pages.CredentialVerifierDashboard,
          displayName: 'Dashboard',
          showInSidebar: true,
        },
        {
          path: '/console/credential-verifier/proofs',
          component: Pages.RequestedProofs,
          displayName: 'Proof Requests',
          showInSidebar: true,
        },
      ],
    },
  ],
});
