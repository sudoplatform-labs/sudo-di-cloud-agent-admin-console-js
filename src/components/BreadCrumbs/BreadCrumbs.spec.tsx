import { Breadcrumb } from 'antd';
import { shallow } from 'enzyme';
import React from 'react';
import { RouteDefinition } from '../../routes';
import { BreadCrumbs } from './BreadCrumbs';

const routes: RouteDefinition[] = [
  {
    path: '/',
    displayName: 'ROOT',
  },
  {
    path: '/a',
  },
  {
    path: '/a/b',
    displayName: 'AN',
  },
  {
    path: '/a/b/c',
    displayName: 'ABC',
  },
];

describe('components/BreadCrumbs', () => {
  describe('BreadCrumbs', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<BreadCrumbs urlPath="/a/b/c" routes={routes} />);
      const renderedBreadcrumbs = wrapper
        .find(Breadcrumb.Item)
        .map((el) => el.key());

      expect(renderedBreadcrumbs).toEqual(['/a/b', '/a/b/c']);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
