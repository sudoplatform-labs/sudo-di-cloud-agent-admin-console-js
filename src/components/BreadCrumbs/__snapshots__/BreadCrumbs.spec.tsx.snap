// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BreadCrumbs BreadCrumbs should render correctly 1`] = `
<Styled(Breadcrumb)
  separator=">"
>
  <BreadcrumbItem
    key="/a/b"
  >
    <Link
      to="/a/b"
    >
      AN
    </Link>
  </BreadcrumbItem>
  <BreadcrumbItem
    key="/a/b/c"
  >
    <Link
      to="/a/b/c"
    >
      ABC
    </Link>
  </BreadcrumbItem>
</Styled(Breadcrumb)>
`;
